File: G_잠꾸러기 수뭉이/Python/Subtasks_2.py
import sys
import heapq
input = sys.stdin.readline

INF = 10**12

# 다익스트라
class _dij:
    class node:
        def __init__(self, nxt, c, l):
            self.nxt = nxt
            self.c = c
            self.l = l

    def __init__(self, n=0):
        self.n = n
        self.adj = [[] for _ in range(n + 1)]
        self.dist = []
        self.pq = []

    def add(self, s, e, c, l):
        self.adj[s].append(self.node(e, c, l))
        self.adj[e].append(self.node(s, c, l))

    def ret(self, s, e, t):
        self.dist = [INF] * (self.n + 1)
        self.pq.clear()
        heapq.heappush(self.pq, (t, s))

        while self.pq:
            cd, cur = heapq.heappop(self.pq)
            if self.dist[cur] <= cd:
                continue
            self.dist[cur] = cd

            for nxt_node in self.adj[cur]:
                nxt, co, ban = nxt_node.nxt, nxt_node.c, nxt_node.l
                if cd >= ban:
                    continue
                if self.dist[nxt] <= cd + co:
                    continue
                heapq.heappush(self.pq, (cd + co, nxt))

        return self.dist[e]

dij = _dij()
t0 = 0
N = 0
M = 0

def decision(cur):
    # 출발 시간이 cur일 때 정점 N까지의 최단 거리
    ret = dij.ret(1, N, cur)
    # 이동 시간이 같으면 True 아니면 False 반환
    return ret <= t0 + cur

def run():
    global N, M, t0, dij
    N, M = map(int, input().split())
    dij = _dij(N)
    # 간선 입력
    for _ in range(M):
        s, e, c, l = map(int, input().split())
        dij.add(s, e, c, l)

    # 출발 시간이 0일 때의 최단 거리
    t0 = dij.ret(1, N, 0)

    # 출발 시간이 0일 때 도달 불가능하면 -1
    if t0 >= INF:
        print(-1)
    else:
        print(t0, end=' ')
        for i in range(21):
            # 출발 시간이 i일 때 이동 시간이 같으면 건너뜀
            if decision(i):
                continue
            # 가장 처음으로 이동 시간이 달라진 출발 시간 출력
            print(i - 1)
            return
        # for문을 다 돌아도 이동 시간이 같으면 20
        print(20)

t = int(input())
for _ in range(t):
    run()

File: G_잠꾸러기 수뭉이/C++/Subtasks_2.cpp
#include <bits/stdc++.h>
using namespace std; using ll = long long;
#define fastio cin.tie(0), cout.tie(0), ios::sync_with_stdio(0);
using pll = pair<ll, ll>;

const ll INF = 1e12;

// 다익스트라
class _dij{
public:
    class node{
    public:
        ll nxt, c, l;
    };

    vector <vector<node>> adj;
    vector <ll> dist;
    priority_queue <pll, vector<pll>, greater<pll>> pq;
    ll n = 0;

    _dij(){}
    _dij(ll n){
        this->n = n;
        adj.clear(); adj.resize(n + 1);
    }

    // 간선 추가
    void add(ll s, ll e, ll c, ll l){
        adj[s].push_back({e, c, l});
        adj[e].push_back({s, c, l});
    }

    // s -> e 최단거리 반환
    ll ret(ll s, ll e, ll t){
        dist.clear(); dist.resize(n + 1, INF);
        while(!pq.empty()) pq.pop();
        pq.push({t, s});

        while(!pq.empty()){
            auto [cd, cur] = pq.top(); pq.pop();
            if(dist[cur] <= cd) continue;
            dist[cur] = cd;

            for(auto&[nxt, co, ban] : adj[cur]){
                // 현재 최단거리가 막차 시간 이상이면 건너 뜀
                if(cd >= ban) continue;
                if(dist[nxt] <= cd + co) continue;
                pq.push({cd + co, nxt});
            }
        } 

        return dist[e];
    }
}; _dij dij;

ll t0, N, M;

bool decision(ll cur){
    // 출발 시간이 cur일 때 정점 N까지의 최단 거리
    ll ret = dij.ret(1, N, cur);

    // 이동 시간이 같으면 1 아니면 0 반환
    return ret <= t0 + cur;
}

void run(){
    cin >> N >> M; dij = {N};
    while(M--){
        // 간선 입력
        ll s, e, c, l; cin >> s >> e >> c >> l;
        dij.add(s, e, c, l);
    }
    
    // 출발 시간이 0일 때의 최단 거리
    t0 = dij.ret(1, N, 0);

    // 출발 시간이 0일 때 정점 N에 도착하지 못하면 -1
    if(t0 >= INF) cout << -1 << "\n"; 
    else{
        cout << t0 << " ";
        for(int i = 0;i <= 20;i++){
            // 출발 시간이 i일 때 출발 시간 0과
            // 이동 시간이 같으면 건너 뜀
            if(decision(i)) continue;

            // 가장 처음으로 이동 시간이 달라진 출발 시간 출력
            cout << i - 1 << "\n"; return;
        }

        // for문을 다 돌아도 이동 시간이 같으면 20
        cout << 20 << "\n";
    }
}

int main(){
    fastio;
    ll t; cin >> t;
    while(t--) run();
    return 0;
}
